cosmosis/samplers/cosmopower_train/cosmopower_train.py

```diff
@@ -0,0 +1,206 @@
+                pickle.dump(reference_prediction, f)
+            
+            # Initiate and train the neural network
+            cp_nn = cosmopower_NN(parameters=list(params_dict_train.keys()),
+                                modes=runs_dict[run]['k_h'][0],
+                                n_hidden=[1048, 1048, 1048, 1048],
+                                verbose=True)
+            
+            training_features = np.log10(runs_dict[run]['p_k']) - np.log10(reference_prediction)
+            with tf.device(self.device):
+                # The training choices are for now hardcoded. These showed to perform well for any CAMB outputs.
+                cp_nn.train(training_parameters=params_dict_train,
+                            training_features=training_features,
+                            filename_saved_model=f'{self.save_dir}_{run}',
+                            validation_split=0.1,
+                            learning_rates=[1e-2, 1e-3, 1e-4, 1e-5, 1e-6, 1e-7],
+                            batch_sizes=[1000, 1000, 1000, 1000, 1000, 1000],
+                            gradient_accumulation_steps=[1, 1, 1, 1, 1, 1],
+                            patience_values=[100, 100, 100, 100, 100, 100],
+                            max_epochs=[1000, 1000, 1000, 1000, 1000, 1000])
+        
+        def run_testing_pk(run, params_dict_test, test_dict, mean_run):
+            reference_prediction = np.squeeze(mean_run.block[run, 'p_k'])
+            # Enter testing phase
+            cp_nn_test = cosmopower_NN(restore=True, restore_filename=f'{self.save_dir}_{run}')
+            emulated_test_spectra = 10.0**(cp_nn_test.predictions_np(params_dict_test) + np.log10(reference_prediction))
+
+            diff = np.abs((emulated_test_spectra - test_dict[run]['p_k']) / test_dict[run]['p_k'])
+            # Define percentiles to calculate
+            percentile_values = [68, 95, 99, 99.9]
+            percentiles = np.array([np.percentile(diff, p, axis=0) for p in percentile_values])
+
+            # Define colors and labels for plotting
+            plot_settings = [
+                {'color': 'salmon', 'label': '99%', 'alpha': 0.8},
+                {'color': 'red', 'label': '95%', 'alpha': 0.7},
+                {'color': 'darkred', 'label': '68%', 'alpha': 1}
+            ]
+
+            plt.figure(figsize=(12, 9))
+            for i, settings in enumerate(plot_settings[::-1]):
+                plt.fill_between(runs_dict[run]['k_h'][0], 0, percentiles[i], **settings)
+
+            plt.xscale('log')
+            plt.legend(frameon=False, fontsize=30, loc='upper left')
+            plt.ylabel(r'$\frac{\vert P(k,z)^{\mathrm{emulated}}_\mathrm{lin} -P(k,z)^{\mathrm{true}}_\mathrm{lin} \vert} {P(k,z)^{\mathrm{true}}_\mathrm{lin} }$', fontsize=30)
+            plt.xlabel(r'$k [h Mpc^{-1}]$', fontsize=30)
+
+            # Save the plot
+            plt.savefig(f'{self.save_dir}_{run}_test.jpg', dpi=200, bbox_inches='tight')
```

Comment:
It's better to save diagrams like these as PNG than JPG. Better resolution. (JPG stores the image in Fourier space, which is great for photos but less good for diagrams with sharp edges).